Rewrite Links in HTML #25

rofe opened the pull request:
Implements #25

trieloff commented on src/utils/mdast-to-vdom.js:
```diff
@@ -61,6 +62,7 @@ class VDOMTransformer {
         return true;
       });
     this._handlers.image = image(options);
+    this._handlers.embed = link(options);
```
Typo: should be `this._handlers.link =…`

trieloff replied:
Btw. I'm unclear why your tests are passing, given that you register the wrong handler.

rofe replied:
Fixed in https://github.com/adobe/hypermedia-pipeline/pull/120/commits/6386202b2e004517b064f2ed2991b8dded8739e9

trieloff commented on src/utils/link-handler.js:
```diff
@@ -0,0 +1,26 @@
+  return function handler(h, node) {
+    const props = {};
+    if (node.url) {
+      props.href = node.url.replace(/\.md/, '.html');
```
It would be ideal to make the extension configurable and dependent on the extension in the request. This might require a bit of mangling as `VDOMTransformer` does not get the current request object, but I think it would increase reusability.

trieloff commented on src/utils/link-handler.js:
```diff
@@ -0,0 +1,26 @@
+    if (node.url) {
+      props.href = node.url.replace(/\.md/, '.html');
+    }
+    if (LINK_TITLES) {
```
I'd simply check if 'node.title' is set. Making the href the title does not seem like a good idea.

rofe replied:
Fixed in https://github.com/adobe/hypermedia-pipeline/pull/120/commits/6386202b2e004517b064f2ed2991b8dded8739e9

trieloff commented on src/utils/link-handler.js:
```diff
@@ -0,0 +1,26 @@
+function link({ LINK_TITLES }) {
+  return function handler(h, node) {
+    const props = {};
+    if (node.url) {
```
You are also rewriting external links, which does not seem like a good idea.

rofe replied:
Fixed in https://github.com/adobe/hypermedia-pipeline/pull/120/commits/6386202b2e004517b064f2ed2991b8dded8739e9

rofe commented:
Addressed several issues from initial commit, such as regressions (empty `<a>` tags), rewritten external links and several other problems. I also removed the title config (needs more thinking).

I tested this version with [Adobe-Developer-Site](https://github.com/adobe/Adobe-Developer-Site) and found no more regressions there. All tests are passing (except one which I commented out, see below):

TODOs:
- uncomment test "Custom Text Matcher with VDOM Nodes" in `testMdastToVDOM.js` - currently fails because the customized link nodes get overwritten by the new link handler

trieloff commented on package.json:
```diff
@@ -38,6 +38,7 @@
     "@adobe/openwhisk-loggly-wrapper": "^0.3.0",
     "ajv": "^6.5.4",
     "callsites": "^3.0.0",
+    "clone": "^2.1.2",
```
Not needed: you can use `Object.assign({}, node)` instead.

rofe replied:
I chose `clone` over `JSON.parse(JSON.stringify(obj))` or `Object.assign({}, obj)` because of potential problems with complex objects (such as `Date`)... but if in a node all we expect to be cloned are strings, numbers and simple objects, we can drop it again.

rofe replied:
Fixed in https://github.com/adobe/hypermedia-pipeline/pull/120/commits/f74f8695efaaaa0ca1ae80cdbbae6cccf81ab4bb

trieloff commented on src/utils/link-handler.js:
```diff
@@ -10,16 +10,18 @@
+  return function handler(h, node, children) {
+    const n = clone(node);
+    if (!url.parse(n.url).protocol) {
+      n.url = n.url.replace(/\.md/, '.html');
```
Maybe just replace it in the path.

trieloff replied:
You never know when someone registers the `.md` top level domain.

rofe replied:
good point

rofe replied:
Fixed in https://github.com/adobe/hypermedia-pipeline/pull/120/commits/5f705848fe86fd2f1726d7b98ed1ce07734a5b0d

trieloff commented on package.json:
```diff
@@ -65,6 +66,7 @@
     "unist-util-map": "^1.0.4",
     "unist-util-select": "^2.0.0",
     "uri-js": "^4.2.2",
+    "url": "^0.11.0",
```
Try to use `uri-js` instead of `url`.

rofe replied:
I chose `url` because `image-handler.js` was already using it.

rofe replied:
Fixed in https://github.com/adobe/hypermedia-pipeline/pull/120/commits/db14b2c6e01f917f1d116c4e419980003569abd2

rofe commented:
last TODOs addressed with https://github.com/adobe/hypermedia-pipeline/pull/120/commits/d968da4d8dc99196ea87e7727ad7ba0a66959104
